src/tools/r2rdump/R2RReader.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.IO;
+using System.Reflection.PortableExecutable;
+
+namespace R2RDump
+{
+    class R2RReader
+    {
+        byte[] pe;
+
+        public string Filename { get; }
+        public bool IsR2R { get; }
+        public Machine Machine { get; }
+        public ulong ImageBase { get; }
+        public R2RHeader R2RHeader { get; }
+
+        public unsafe R2RReader(string filename)
+        {
+            Filename = filename;
+            pe = File.ReadAllBytes(filename);
+
+            fixed (byte* p = pe)
+            {
+                IntPtr ptr = (IntPtr)p;
+                PEReader peReader = new PEReader(p, pe.Length);
```

Comment:
@jkotas @nattress Previously, the PEReader reads the image from the filestream and my program reads the image again to get the byte[]. I made some changes so it only reads once, but it uses unsafe. Would that be ok?

Reply:
I would consider using the `public PEReader(ImmutableArray<byte> peImage)` constructor on `PEReader`. This would let you pass around an `ImmutableArray<byte>` instead of the raw `byte[]`, preventing us from accidentally modifying the file's bytes. We'd no longer need unsafe. Creating the ImmutableArray from the byte[] returned by `File.ReadAllBytes` will cause an array copy but I think it's worth the safety guarantee.

Reply:
(This can be in a follow-up PR by the way)

Reply:
How about `ReadOnlySpan<byte>`? 

Seems like spans are ideally suited for avoiding needless copying of image file bits.